kafka-avro-adapter/src/cs_producer.cpp

```diff
@@ -18,140 +18,177 @@
+    if (avro_value_get_type(value) == AVRO_RECORD)
     {
-        avro_value_t field;
-        avro_value_get_by_index(value, i, &field, NULL);
-        const char* name = avro_schema_record_field_name(avro_value_get_schema(value), i);
-        int table_idx = table.getColumn(name).getPosition();
+        size_t n_fields = 0;
+        avro_value_get_size(value, &n_fields);
 
-        switch (avro_value_get_type(&field))
+        for (size_t i = 0; i < n_fields; i++)
         {
-            case AVRO_STRING:
-            {
-                const char* s;
-                size_t len;
-                int r = avro_value_get_string(&field, &s, &len);
-                assert(r == 0);
-                std::string str = std::string(s, len);
-                logger() << "STRING: " << str << std::endl;
-                insert->setColumn(table_idx, str);
-                break;
-            }
-            case AVRO_BYTES:
-            {
-                const void* s;
-                size_t len;
-                int r = avro_value_get_bytes(&field, &s, &len);
-                assert(r == 0);
-                insert->setColumn(table_idx, std::string((char*)s, len));
+            avro_value_t field;
+            avro_value_get_by_index(value, i, &field, NULL);
+            const char* name = avro_schema_record_field_name(avro_value_get_schema(value), i);
+            int table_idx = table.getColumn(name).getPosition();
 
-                break;
-            }
-            case AVRO_INT32:
-            {
-                int32_t out;
-                int r = avro_value_get_int(&field, &out);
-                assert(r == 0);
-                insert->setColumn(table_idx, out);
-                break;
-            }
-            case AVRO_INT64:
-            {
-                int64_t out;
-                int r = avro_value_get_long(&field, &out);
-                assert(r == 0);
-                insert->setColumn(table_idx, out);
-                break;
-            }
-            case AVRO_FLOAT:
-            {
-                float out;
-                int r = avro_value_get_float(&field, &out);
-                assert(r == 0);
-                insert->setColumn(table_idx, out);
-                break;
-            }
-            case AVRO_DOUBLE:
-            {
-                double out;
-                int r = avro_value_get_double(&field, &out);
-                assert(r == 0);
-                insert->setColumn(table_idx, out);
-                break;
-            }
-            case AVRO_BOOLEAN:
+            if ( avro_value_get_type(&field) == AVRO_UNION )
             {
-                int out = 0;
-                int r = avro_value_get_boolean(&field, &out);
-                assert(r == 0);
-                insert->setColumn(table_idx, (bool)out);
-                break;
+                int disc;
+                avro_value_get_discriminant(&field, &disc);
+         
+                if ( disc > 0 )
```

Comment:
This assumes that the first element of the UNION is null but the Avro specification doesn't require it (it merely suggests using it). Does the data source where this is read from always have the first type in an UNION as a null? If so, it would be good to add an assertion here.